File: Leetcode/Count Vowels Permutation/Count Vowels Permutation.py

#Problem :# https://leetcode.com/problems/count-vowels-permutation/
class Solution(object):
    def countVowelPermutation(self, n):

        counts = {vowel: 1 for vowel in "aeiou"}    # 1 permutation for each letter

        # Create a mapping from each vowel to the number of permutations ending with that vowel.
        # For each new letter, update the permutations from all previous vowels that can be followed by the new vowel.
        for _ in range(n - 1):
            new_counts = {}
            new_counts["a"] = counts["e"] + counts["i"] + counts["u"]
            new_counts["e"] = counts["a"] + counts["i"]
            new_counts["i"] = counts["e"] + counts["o"]
            new_counts["o"] = counts["i"]
            new_counts["u"] = counts["i"] + counts["o"]
            counts = new_counts

        return sum(counts.values()) % (10 ** 9 + 7)

File: Leetcode/Count Vowels Permutation/Count Vowels Permutation.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
  int countVowelPermutation(int n) 
  {
        //returning 5 if n is equal to 1 
        if(n==1)
        {
            return 5;
        }
      
        //variable declaration
        long a=1,e=1,i=1,o=1,u=1,m=1000000007;
      
        //loop for finding vowel permutations 
        for(int ii=1;ii<n;ii++)
        {
            long na=(e+i+u)%m, ne=(a+i)%m, ni=(e+o)%m, no=i, nu=(i+o)%m;
            a=na,e=ne,i=ni,o=no,u=nu;
        }
      
        //returning the count
        return (a+e+i+o+u)%m;
    }
};